internal/controller/nicdevice_controller.go

```diff
@@ -0,0 +1,550 @@
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// applyRuntimeConfig applies each device's runtime spec in parallel
+// if update is successful, applies status condition UpdateSuccessful, otherwise RuntimeConfigUpdateFailed
+// if rebootRequired, sets status condition PendingReboot
+// if status.rebootRequired == true, skips the device
+// returns nil if all devices' config updates were successful, error otherwise
+func (r *NicDeviceReconciler) applyRuntimeConfig(ctx context.Context, statuses nicDeviceConfigurationStatuses) error {
+	var wg sync.WaitGroup
+
+	for i := 0; i < len(statuses); i++ {
+		wg.Add(1)
+		go func(index int) {
+			defer wg.Done()
+
+			status := statuses[i]
+			status.lastStageError = nil
+			if status.rebootRequired {
+				return
+			}
+
+			lastAppliedState, found := status.device.Annotations[consts.LastAppliedStateAnnotation]
+			if found {
+				specJson, err := json.Marshal(status.device.Spec)
+				if err != nil {
+					status.lastStageError = err
+					return
+				}
+
+				if string(specJson) != lastAppliedState {
+					log.Log.V(2).Info("last applied state differs, reboot required", "device", status.device.Name)
+					status.rebootRequired = true
+
+					err := r.updateDeviceStatusCondition(ctx, status.device, consts.PendingRebootReason, metav1.ConditionTrue, "")
+					if err != nil {
+						status.lastStageError = err
+						return
+					}
+
+					return
+				}
+			}
+
+			err := r.hostManager.ApplyDeviceRuntimeSpec(statuses[i].device)
+			if err != nil {
+				statuses[i].lastStageError = err
+				err = r.updateDeviceStatusCondition(ctx, status.device, consts.RuntimeConfigUpdateFailedReason, metav1.ConditionFalse, err.Error())
+				return
+			}
+
+			specJson, err := json.Marshal(status.device.Spec)
+			if err != nil {
+				status.lastStageError = err
+				return
+			}
+
+			if status.device.Annotations == nil {
+				status.device.SetAnnotations(make(map[string]string))
+			}
+			status.device.Annotations[consts.LastAppliedStateAnnotation] = string(specJson)
+			err = r.Update(ctx, status.device)
+			if err != nil {
+				status.lastStageError = err
+				return
+			}
+
+			err = r.updateDeviceStatusCondition(ctx, status.device, consts.UpdateSuccessfulReason, metav1.ConditionFalse, "")
+			if err != nil {
+				status.lastStageError = err
+				return
+			}
+		}(i)
+	}
+
+	wg.Wait()
+
+	for _, status := range statuses {
+		if status.lastStageError != nil {
+			return status.lastStageError
+		}
+	}
+
+	return nil
+}
+
+// handleReboot schedules maintenance and reboots the node if maintenance is allowed
+// Before rebooting the node, strips LastAppliedState annotations from all devices
+// returns true if requeue of the reconcile request is required, false otherwise
+// return err if encountered an error while performing maintenance scheduling / reboot
+func (r *NicDeviceReconciler) handleReboot(ctx context.Context, statuses nicDeviceConfigurationStatuses) (bool, error) {
```

Comment:
I think it's better to return with a controller result (i.e., `ctrl.Result{RequeueAfter: requeueTime`) instead of a boolean. The boolean makes it harder to understand what's happening. `handleReboot` returns with `true` which usually indicates the the reboot has been handled